pkg/auth/dependency/webapp/service.go

```diff
@@ -45,26 +50,61 @@ func NewServiceLogger(lf *log.Factory) ServiceLogger {
 	return ServiceLogger{lf.New("webapp-service")}
 }
 
+const UserAgentTokenCookieName = "ua-token"
+
 type Service struct {
-	Logger ServiceLogger
-	Store  Store
-	Graph  GraphService
+	Logger       ServiceLogger
+	Request      *http.Request
+	Store        Store
+	Graph        GraphService
+	ServerConfig *config.ServerConfig
 }
 
-func (s *Service) GetState(stateID string) (state *State, err error) {
-	if stateID != "" {
-		state, err = s.Store.Get(stateID)
-		if err != nil {
-			return
-		}
+func (s *Service) getUserAgentToken() string {
+	token, err := s.Request.Cookie(UserAgentTokenCookieName)
+	if err != nil {
+		return ""
 	}
-	return
+	return token.Value
+}
+
+func (s *Service) generateUserAgentToken() (string, *http.Cookie) {
+	token := corerand.StringWithAlphabet(32, base32.Alphabet, corerand.SecureRand)
+	secure := httputil.GetProto(s.Request, s.ServerConfig.TrustProxy) == "https"
+
+	cookie := &http.Cookie{
+		Name:     UserAgentTokenCookieName,
+		Value:    token,
+		Path:     "/",
+		Secure:   secure,
+		MaxAge:   0,                     // Use HTTP session cookie; expires when browser closes
+		SameSite: http.SameSiteNoneMode, // Ensure resume-able after redirecting from external site
+	}
+	if !samesite.ShouldSendSameSiteNone(s.Request.UserAgent(), secure) {
+		cookie.SameSite = 0
+	}
```

Comment:
It seems that we now have more places to create cookie. Maybe it is a good chance to introduce a CookieMaker which take care of setting the correct Secure and SameSite flag.

My proposal

```golang
type CookieMaker interface {
  // NewCookie returns a pointer to http.Cookie with Secure and SameSite set to proper values.
  // All other values are left untouched. You must initialize them before further use.
  MakeCookie() *http.Cookie
}
```